packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+}
+
+/// Returns a vector of the component indices of all rendered children that are elements.
+///
+/// If the component extends another component, the children from the extend source come first.
+fn get_non_string_rendered_children_including_from_extend(
+    component_idx: ComponentIdx,
+    components: &Vec<Rc<RefCell<ComponentEnum>>>,
+) -> Vec<ComponentIdx> {
+    // TODO: the behavior of this function should mirror the .to_flat_dast method
+    // of the RenderedComponent trait.
+    // Is there a way to specify the behavior in one place so they cannot diverge and cause bugs?
+    // Perhaps, components cannot override the whole .to_flat_dast method, but just pieces
+    // such as determining the data sent.
+    // And, if they override the behavior is should be through methods called in both places,
+    // like the .get_rendered_children method.
+
+    let component = components[component_idx].borrow_mut();
+
+    let mut children = if let Some(&ExtendSource::Component(source_idx)) = component.get_extend() {
+        get_non_string_rendered_children_including_from_extend(source_idx, components)
+    } else {
+        Vec::new()
+    };
+
+    let mut children2 = component
+        .get_rendered_children()
+        .iter()
+        .filter_map(|child| match child {
+            &ComponentChild::Component(comp_idx) => Some(comp_idx),
+            _ => None,
+        })
+        .collect();
+
+    children.append(&mut children2);
+
+    children
+}
+
+/// If the state variable specified by original_component_state is stale or unresolved,
+/// then freshen the variable, resolving its dependencies if necessary.
+///
+/// If the state variable was not fresh, then recurse to its dependencies to freshen them.
+pub fn freshen_state_var(
+    original_component_state: ComponentStateDescription,
+    components: &Vec<Rc<RefCell<ComponentEnum>>>,
+    dependencies: &mut Vec<Vec<Vec<Vec<Dependency>>>>,
+    dependent_on_state_var: &mut Vec<Vec<Vec<ComponentStateDescription>>>,
+    dependent_on_essential: &mut Vec<HashMap<EssentialDataOrigin, Vec<ComponentStateDescription>>>,
+    essential_data: &mut Vec<HashMap<EssentialDataOrigin, EssentialStateVar>>,
+    freshen_stack: &mut Vec<StateVarCalculationState>,
+    should_initialize_essential_data: bool,
+) -> () {
+    // This function currently implements recursion through an iterative method,
+    // using a stack on the heap.
+    // This approach was chosen because the function recursion implementation would overflow
+    // the small WASM stack once it recursed a few thousands of times.
+    // An alternative approach would be to increase the size of the WASM stack.
+    // This approach was chosen because it is relatively easy to produce documents
+    // with thousands of levels in the dependency graph, and it wasn't clear what
+    // size WASM stack would be appropriate.
+
+    let current_freshness = components[original_component_state.component_idx]
+        .borrow()
```

Comment:
We could do that. Looking through the code, this seems to be the only place where we want to find the freshness by state variable index, so I'm not sure that it is worth it (given that it would involve the derive macro to pipe it through the enum).